### PR TITLE
Relate to keystone to get credentials for exporter

### PR DESCRIPTION
**Pull request by samuelallan72**
Save these credentials to a clouds.yaml file,
so they can be used by openstack-exporter.

If keystone in your test environment uses https, you may need to manually set the `ssl_ca` config.

internal ticket: [BSENG-2172](https://warthogs.atlassian.net/browse/BSENG-2172)

[BSENG-2172]: https://warthogs.atlassian.net/browse/BSENG-2172?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -9,19 +9,132 @@
...
         """Initialize the charm."""
         super().__init__(*args)
+        self.framework.observe(self.on.upgrade_charm, self._on_upgrade)
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.credentials_relation_changed, self._on_credentials_changed)
+        self.framework.observe(self.on.collect_unit_status, self._on_collect_unit_status)
+
+    def _on_upgrade(self, _: UpgradeCharmEvent) -> None:
+        """Reconfigure everything on upgrade."""
+        self._configure()
```
Comment (chanchiwai-ray) [inline src/charm.py:40]: This is unnecessary, please remove it. Charm upgrade will run config event https://juju.is/docs/sdk/upgrade-charm-event

File: src/charm.py [hunk 2]
```diff
@@ -9,19 +9,132 @@
...
+        https://docs.openstack.org/keystone/latest/contributor/http-api.html
+        """
+        auth_url = "{protocol}://{hostname}:{port}/v3".format(
+            protocol=data["service_protocol"],
+            hostname=data["service_hostname"],
+            port=data["service_port"],
+        )
+        contents = {
+            "clouds": {
+                CLOUD_NAME: {  # the name here must be passed to --cloud option of openstack-exporter
```
Comment (chanchiwai-ray) [inline src/charm.py:96]: This is not necessary comment

**Review by chanchiwai-ray — CHANGES_REQUESTED** [review]
Small issues and some questions

File: README.md [hunk 3]
```diff
@@ -7,3 +7,21 @@ the Prometheus monitoring system. This operator works with the Canonical
 Observability Stack (COS).
 
 More information: https://charmhub.io/openstack-exporter-operator
+
+## Development
```
[the remark this answers is not in the document]
Comment (chanchiwai-ray) [reply README.md:11]: A link to `CONTRIBUTION.md` would be nice

File: src/charm.py (hunk 2 above)
Comment (chanchiwai-ray) [reply src/charm.py:91]: well, at least these will be hardcoded in the charm anyways

File: README.md [hunk 4]
```diff
@@ -7,3 +7,5 @@ the Prometheus monitoring system. This operator works with the Canonical
 Observability Stack (COS).
 
 More information: https://charmhub.io/openstack-exporter-operator
+
+For developer documentation, please see [./CONTRIBUTING.md](./CONTRIBUTING.md).
```
Comment (jneo8) [inline README.md:11]: **nitpick (non-blocking):** 

Consider move this to another independent PR.

File: charmcraft.yaml [hunk 5]
```diff
@@ -16,13 +16,17 @@ bases:
...
+  credentials:
+    interface: keystone-admin
+
+config:
+  options:
+    ssl_ca:
+      default:
+      type: string
+      description: |
+        Custom SSL CA for keystone if required.
```
Comment (chanchiwai-ray) [inline charmcraft.yaml:29]: Sorry, one final spot: we should probably document this a bit more (it's a base64 encoded string or plain multi-line string (because it's a CA)). Users coming from the old [charm](https://git.launchpad.net/charm-prometheus-openstack-exporter/tree/src/reactive/openstack_exporter.py#n192) will assume that it's base64 encoded. For me, multi-line string is kind of hard to use

File: src/charm.py [hunk 6]
```diff
@@ -9,19 +9,127 @@
...
+                "service_username",
+                "service_password",
+                "service_project_name",
+                "service_project_domain_name",
+                "service_user_domain_name",
+                "service_region",
+            ]
+        )
+
+    def update_clouds_yaml(self, data: dict[str, str]) -> None:
```
Comment (jneo8) [inline src/charm.py:77]: **nitpick (non-blocking):** 

The current function name does not clearly indicate its purpose. User need background knowledge that the exporter needs a 'clouds.yaml' config file. A more descriptive name like 'generate_exporter_config' might be better

File: src/charm.py [hunk 7]
```diff
@@ -9,19 +9,127 @@
...
+        if data:
+            logger.info("Keystone credentials are available, starting services.")
+            self.update_clouds_yaml(data)
+            # TODO: when openstack-exporter service is added, restart it here
+        else:
+            logger.info("Keystone credentials are not available, stopping services.")
+            # TODO: when openstack-exporter service is added, stop it here
+
+    def _is_keystone_data_ready(self, data: dict[str, str]) -> bool:
+        """Check if all the data is available from keystone."""
```
Comment (jneo8) [inline src/charm.py:61]: **nitpick:** 

You probably want to add doc-string here to explain the meaning of parameter `data`.

**Review by jneo8 — CHANGES_REQUESTED** [review]

File: src/charm.py [hunk 8]
```diff
@@ -9,19 +9,132 @@
...
+                    "auth": {
+                        "username": data["service_username"],
+                        "password": data["service_password"],
+                        "project_name": data["service_project_name"],
+                        "project_domain_name": data["service_project_domain_name"],
+                        "user_domain_name": data["service_user_domain_name"],
+                        "auth_url": auth_url,
+                    },
+                    "verify": data["service_protocol"] == "https",
+                    "cacert": self.config["ssl_ca"] or None,
```
[the remark this answers is not in the document]
Comment (jneo8) [reply src/charm.py:104]: +1 good point from Ray.

Provide through the config is fine now. The behavior is as same as the *charm-prometheus-openstack-exporter*.

I think we can create a issue here as a backlog.

File: charmcraft.yaml (hunk 5 above)
Comment (chanchiwai-ray) [reply charmcraft.yaml:29]: Sure, that's fine. Just want to point that out loudly

File: charmcraft.yaml [hunk 9]
```diff
@@ -16,13 +16,20 @@ bases:
...
-# config:
-#   options:
-#     snap_channel:
-#       default: "latest/stable"
-#       type: string
-#       description: |
-#           If install_method is set to "snap" this option controls channel name.
+requires:
+  credentials:
+    interface: keystone-admin
```
Comment (chanchiwai-ray) [inline charmcraft.yaml:21]: question: what is the difference between keystone-admin and keystone-credential?

Comment (chanchiwai-ray) [reply charmcraft.yaml:21]: oh, side note: I was just wanting to leave a comment, but the status become request-change

File: src/charm.py [hunk 10]
```diff
@@ -9,19 +9,133 @@
...
```
Comment (chanchiwai-ray) [inline src/charm.py:110]: `self.config["ssl_ca"] or None` should be changed to `self.config.get("ssl_ca", "")` since the default value of that config is `unset` which is causes 

```
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-openstack-exporter-3/charm/./src/charm.py", line 240, in <module>
    ops.main(OpenstackExporterOperatorCharm)
  File "/var/lib/juju/agents/unit-openstack-exporter-3/charm/venv/ops/main.py", line 471, in __call__
    return main(charm_class, use_juju_for_storage=use_juju_for_storage)
  File "/var/lib/juju/agents/unit-openstack-exporter-3/charm/venv/ops/main.py", line 454, in main
    framework.reemit()
  File "/var/lib/juju/agents/unit-openstack-exporter-3/charm/venv/ops/framework.py", line 875, in reemit
    self._reemit()
  File "/var/lib/juju/agents/unit-openstack-exporter-3/charm/venv/ops/framework.py", line 955, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-openstack-exporter-3/charm/./src/charm.py", line 210, in _on_config_changed
    self.configure(event)
  File "/var/lib/juju/agents/unit-openstack-exporter-3/charm/./src/charm.py", line 160, in configure
    self.write_cloud_config(data)
  File "/var/lib/juju/agents/unit-openstack-exporter-3/charm/./src/charm.py", line 107, in write_cloud_config
    "cacert": self.config["ssl_ca"] or None,
  File "/var/lib/juju/agents/unit-openstack-exporter-3/charm/venv/ops/model.py", line 796, in __getitem__
    return self._data[key]
KeyError: 'ssl_ca'
```

Comment (chanchiwai-ray) [reply src/charm.py]: oh, why not `self.config.get("ssl_ca")` without `or None`

Comment (chanchiwai-ray) [reply src/charm.py]: I don't get it.. Isn't the one you describing the same as `self.config.get("ssl_ca")`?

If `ssl_ca` is unset, set to null, or empty string, `self.config.get("ssl_ca")` will return `None`

Comment (chanchiwai-ray) [reply src/charm.py]: that also works, thanks!

File: src/charm.py [hunk 11]
```diff
@@ -9,19 +9,133 @@
...
+        This is an idempotent method to configure everything.
+        """
+        data = self._get_keystone_data()
+
+        if data:
+            logger.info("Keystone credentials are available, starting services.")
+            self.write_cloud_config(data)
+            # TODO: when openstack-exporter service is added, restart it here
+        else:
+            logger.info("Keystone credentials are not available, stopping services.")
```
Comment (chanchiwai-ray) [inline src/charm.py:57]: Do we need to defer here? I can imagine if we can't get the relation data the first try when credentials changed, then we need to remove-relation and relate again

**Review by chanchiwai-ray — DISMISSED** [review]
But overall it looks good to me. I don't want to be picky right now and block the PR here

File: src/charm.py (hunk 11 above)
Comment (chanchiwai-ray) [reply src/charm.py:57]: Okay, thanks

File: charmcraft.yaml [hunk 12]
```diff
@@ -16,13 +16,20 @@ bases:
...
+config:
+  options:
+    ssl_ca:
+      default:
+      type: string
+      description: |
+        Custom SSL CA for keystone if required.
+
+        The format should be the contents of a PEM encoded file.
+        (no base64 encoding required).
```
[the remark this answers is not in the document]
Comment (jneo8) [reply charmcraft.yaml:32]: Sure.

[the remark this answers is not in the document]
Comment (jneo8) [reply charmcraft.yaml:32]: #9

**Review by jneo8 — DISMISSED** [review]

File: src/charm.py [hunk 13]
```diff
@@ -35,8 +33,10 @@
 
 
 # Snap config options global constants
+# This is to match between openstack-exporter and the entry in clouds.yaml
 CLOUD_NAME = "openstack"
-OS_CLIENT_CONFIG = USER_HOME / "clouds.yaml"
+# store the clouds.yaml where it's easily accessible by the openstack-exporter snap
+OS_CLIENT_CONFIG = "/var/snap/golang-openstack-exporter/common/clouds.yaml"
```
Comment (chanchiwai-ray) [inline src/charm.py:39]: Nice find, I almost forgot the `SNAP_COMMON` directory

File: src/charm.py [hunk 14]
```diff
@@ -94,8 +166,14 @@ def configure(self, event: ops.HookEvent) -> None:
...
-        # keystone and grafana-agent.
+        data = self._get_keystone_data()
+        if not data:
+            logger.info("Keystone credentials are not available, stopping services.")
+            snap_service.stop()
+            return
+
+        logger.info("Keystone credentials are available, starting services.")
+        self._write_cloud_config(data)
         snap_service.start()
```
Comment (chanchiwai-ray) [inline src/charm.py:179]: We might need a `snap_service.restart()`, I tested manually with the snap: starting an already started service will not take effect

**Review by chanchiwai-ray — APPROVED** [review]
Thanks, overall looks good, I left a comment about `snap_service.restart()`.

(optional and opinionated) Can we make the non-event function without the `_` prefix? It just looks nicer to me, and make me easier to visually distinguish the event callbacks and normal functions

File: src/charm.py (hunk 14 above)
Comment (chanchiwai-ray) [reply src/charm.py:179]: oh, it should have been `changes-requested` because it's kind of a bug

**Comment by chanchiwai-ray** [comment]
> > Can we make the non-event function without the _ prefix? It just looks nicer to me, and make me easier to visually distinguish the event callbacks and normal functions
> 
> So ideally all the other methods should have the `_` prefix too, as this is used as a convention to show that they are internal methods and shouldn't be called from outside the class.

Yes, I understand that, so it's optional

**Review by chanchiwai-ray — APPROVED** [review]
Thanks, LGTM.

Did some basic testing with keystone:

- [x] installed with resource -> Blocked: missing keystone relation
- [x] relate to keystone ->  Active
- [x] curl the endpoint -> Blocked: snap service is not running (after update-status) (note this is expected with only keystone)
- [ ] changing config ssl_ca render the `clouds.yaml`
